includes/class-wc-coupon.php

```diff
@@ -154,131 +319,502 @@ private function get_coupon_id_from_code( $code ) {
+		$this->set_excluded_product_categories( $exclude_product_categories );
+		$exclude_sale_items = ( 'yes' === get_post_meta( $coupon_id, 'exclude_sale_items', true ) );
+		$this->set_exclude_sale_items( $exclude_sale_items );
+		$this->set_minimum_amount( get_post_meta( $coupon_id, 'minimum_amount', true ) );
+		$this->set_maximum_amount( get_post_meta( $coupon_id, 'maximum_amount', true ) );
+		$this->set_email_restrictions( get_post_meta( $coupon_id, 'customer_email', true ) );
+		$this->set_used_by( (array) get_post_meta( $coupon_id, '_used_by' ) );
+
+		// Load custom set metadata (coupon custom fields)
+		$meta_data   = get_post_meta( $coupon_id );
+		$ignore_keys = array_merge( $this->_data, array( '_used_by' => '', 'coupon_amount' => '', '_edit_lock' => '', '_edit_last' => '' ) );
+		$ignore_keys = array_keys( $ignore_keys );
+		$meta_data   = array_diff_key( $meta_data, array_fill_keys( $ignore_keys, '' ) );
+		foreach ( $meta_data as $key => $value ) {
+			$this->_meta_data[ $key ] = $value;
 		}
 
 		do_action( 'woocommerce_coupon_loaded', $this );
 	}
 
 	/**
-	 * Format loaded data as array.
-	 * @param  string|array $array
-	 * @return array
+	 * Create a new coupon.
+	 * @since 2.6.0
 	 */
-	public function format_array( $array ) {
-		if ( ! is_array( $array ) ) {
-			if ( is_serialized( $array ) ) {
-				$array = maybe_unserialize( $array );
-			} else {
-				$array = explode( ',', $array );
-			}
+	public function create() {
+		$coupon_id = wp_insert_post( apply_filters( 'woocommerce_new_coupon_data', array(
+			'post_type'    => 'shop_coupon',
+			'post_status'  => 'publish',
+			'post_author'  => get_current_user_id(),
+			'post_title'   => $this->get_code(),
+			'post_content' => '',
+			'post_excerpt' => $this->get_description(),
+		) ), true );
+
+		if ( $coupon_id ) {
+			$this->_data['id'] = $coupon_id;
+			$this->update_post_meta( $coupon_id );
+			do_action( 'woocommerce_new_coupon', $coupon_id );
 		}
-		return array_filter( array_map( 'trim', array_map( 'strtolower', $array ) ) );
 	}
 
 	/**
-	 * Check if coupon needs applying before tax.
-	 *
-	 * @return bool
+	 * Updates an existing coupon.
+	 * @since 2.6.0
 	 */
-	public function apply_before_tax() {
-		return true;
+	public function update() {
+		$coupon_id = $this->get_id();
+
+		$post_data = array(
+			'ID' => $coupon_id,
+			'post_title'   => $this->get_code(),
+			'post_excerpt' => $this->get_description(),
+		);
+
+		wp_update_post( $post_data );
+		$this->update_post_meta( $coupon_id );
+		do_action( 'woocommerce_update_coupon', $coupon_id );
 	}
 
 	/**
-	 * Check if a coupon enables free shipping.
-	 *
-	 * @return bool
+	 * Save data (either create or update depending on if we are working on an existing coupon)
+	 * @since 2.6.0
 	 */
-	public function enable_free_shipping() {
-		return 'yes' === $this->free_shipping;
+	public function save() {
```

Comment:
Save should return `$this->get_id()` to match what orders are doing in the order CRUD branch.